Create microsite

mtsokol opened the pull request:
#15 
@mijicd Hi, I've just completed first draft of mdoc microsite and I would be grateful for a review and advice.
While working on this part I have encountered few issues which I would like to ask about.


So the general design is based on `ZIO`'s microsite. The `Essentials` are similar to `Data Types` where basic API usage is shown so it's going to be about library design. 

For now there are introduction page where the NIO idea is repeated and some references,
page about `SocketChannel`s and page about `FileChannel`s where simple operations are shown.


The `Use Cases` show usage patterns in more complex cases.

The one will be the already existing example of `Client-Server`.

For the second example, as you suggested, I've used `zio-actor`. I've decided to implement `actor-remote` module, just like basic `akka-remote`.

There will maybe appear something more for each section.

---
###  Questions that arose

1. Is such microsite ok? As mentioned I've followed `ZIO` microsite

2. For `Essentials` which data types should be also explained? `AsynchronousChannelGroup` or `Pipe` or `Selector`?

3. With `Actor Remoting` I've encountered few issues - right now the example does not compile:
- First because the snippets from `zio-actor` test suite fail to compile (!) https://github.com/zio/zio-actors/issues/47 (I'm using latest maven artifact: `0.0.1+10-51dde802`)
- Second the latest maven `zio-actor` artifact is way behind master because it uses `scalaz.zio` instead of `zio.zio` and the `for` comprehensions fail to compile due to different monads used.

How to deal with it? Build the master on my own or wait until new release?

4. Also in `zio-nio` on `test` none is executed and neither with expilcite `test:runMain` command  

5. Once again about this `Actor Remoting` example. What do you think about attempt to initialize such module for `zio` ecosystem? like `zio-remote` similar to `akka-remote` where such functionality could be implemented? 
---

As I said it's WIP but it builds via `docs/docusaurusCreateSite` and `yarn start` after `cd /website`

mijicd commented:
@mtsokol First of all, thanks for the contribution! I think you misunderstood my suggestion. When I mentioned `zio-actors` I meant their docs, not implementing the functionality (you're more than welcome to grab [this one](https://github.com/zio/zio-actors/issues/19)). Thus being said, starting the review :).

mijicd commented on build.sbt:
```diff
@@ -36,3 +36,19 @@ lazy val zioNio = project
+    scalacOptions -= "-Xfatal-warnings",
+    libraryDependencies ++= Seq(
+      "dev.zio"      %% "zio"        % "1.0.0-RC11-1",
+      "dev.zio"      %% "zio-actors" % "0.0.1+10-51dde802",
```
As said in the comment, `zio-actors` were pointed for the sake of documentation, no need to implement remoting here.

mtsokol replied:
Sure, removed

mijicd commented on docs/essentials/files.md:
````diff
@@ -0,0 +1,57 @@
+
+```scala mdoc:silent
+val channelM = for {
+  path <- ZIO.succeedLazy(Paths.get("./file.txt"))
````
`succeedLazy` is about to be deprecated from `ZIO`, therefore I'd argue that we shouldn't include it in examples.

mtsokol replied:
Changed to `effectTotal`

mijicd commented on docs/essentials/files.md:
```diff
@@ -0,0 +1,57 @@
+    
+  fileLock <- channel.lock(position = 0, size = 10, shared = false)
+  _ = fileLock.overlaps(5, 20)               //true
+  _ = fileLock.close()                       //same as release
```
Side note: we should manage resources via `ZManaged`. As far as I'm aware of, this issue hasn't been migrated from old repo yet.

@quelgar @pshemass ^

mtsokol replied:
Sure, as a resource it should be treated accordingly. 
Here I refactored it by using `bracket` (credits for @sinoz for showing it here #22) in the first part and `Managed` in the second part, for demonstration purposes.

mijicd commented on docs/essentials/index.md:
```diff
@@ -0,0 +1,34 @@
+ 
+## So what is NIO once again? 
+
+As standard input-output processing we are used to stream-like operations. 
```
This section sounds a bit unstructured. Personally, I'd even consider skipping it and just referencing the relevant Javadocs, or keeping it shorter.

mtsokol replied:
Sure, removed it and added nio javadocs link. It's not enough for clear comprehensive explanation and making it shorter won't help either.

mijicd commented on docs/usecases/actor_remoting.md:
```diff
@@ -0,0 +1,204 @@
+---
+id: usecases_actor_remoting
+title:  "Actor Remoting"
```
This is amazing effort, but as said, it was caused by misunderstanding. I propose that apply the idea [here](https://github.com/zio/zio-actors/issues/19), and to provide some simpler examples here.

mtsokol replied:
I've misunderstood it but I was also thinking about writing such module for `zio-actors`. I will take this piece to present this idea in issue or gitter.

But still I will provide some other examples there.

mijicd replied:
It would be very useful, and there's already an issue, i've referenced it in my comment, so take a look!

mtsokol commented:
@mijicd I've addressed comments, also added that circleci mdoc job (with scaladocs generation). 
Going back to my `2.` question, which data types are also worth describing in `Essentials`?

quelgar commented:
This is great! Your "Essentials" covers files and sockets, which I think are indeed the most important aspects. Good to go IMHO.

@mijicd great point about using `ZManaged`. Really we should be aiming to not have any `close` methods, correct?

mijicd commented:
@quelgar That would be ideal. I'm willing to take that one on me, as soon as we have an open ticket :)

jczuchnowski commented on website/core/Footer.js:
```diff
@@ -0,0 +1,65 @@
+              <h5>Additional resources</h5>
+              <a
+                  href="https://javadoc.io/doc/dev.zio/zio_2.12/">
+                Scaladoc of zio
```
make uppercase ZIO

mtsokol replied:
Sure, corrected it

jczuchnowski commented on docs/usecases/client_server.md:
```diff
@@ -0,0 +1,6 @@
+title:  "Client Server"
+---
+
+TODO remake existing `channel.md`
```
So what do we need here?

mtsokol replied:
I've moved here the existing `tut/channel.md`, split it into functions, added comment for each part and provided example output. Also I've removed the old one.

jczuchnowski commented:
Looks good. I think we should merge it as fast as possible to enable more contributions to the documentation.
